#! /bin/sh /usr/share/dpatch/dpatch-run
## fix_xen.dpatch by Andrew Deason <adeason2@uiuc.edu>
##
## All lines beginning with `## DP:' are a description of the patch.
## DP: Fixes Xen-specific build options by actually passing CFLAGS to
## DP: libgc's Makefile

@DPATCH@

--- mono/configure.in	2008-05-19 21:13:13.000000000 -0500
+++ mono/configure.in	2008-05-19 21:13:21.000000000 -0500
@@ -36,6 +36,11 @@
   gc_default=boehm
 fi
 
+# These variables are the CPPFLAGS/CFLAGS passed to libgc's configure
+# libgc should inherit the original CFLAGS/CPPFLAGS passed to configure, i.e. -O0
+CPPFLAGS_FOR_LIBGC=$CPPFLAGS
+CFLAGS_FOR_LIBGC=$CFLAGS
+
 #
 # These are the flags that need to be stored in the mono.pc file for 
 # compiling code that will embed Mono
@@ -301,6 +306,8 @@
 #AC_DISABLE_FAST_INSTALL
 AM_PROG_LIBTOOL
 
+DOLT
+
 # Test whenever ld supports -version-script
 AC_PROG_LD
 AC_PROG_LD_GNU
@@ -399,6 +406,7 @@
 	esac
 fi
 CFLAGS="$CFLAGS -g $WARN"
+CFLAGS_FOR_LIBGC="$CFLAGS_FOR_LIBGC -g"
 
 # Where's the 'mcs' source tree?
 if test -d $srcdir/mcs; then
@@ -556,6 +564,8 @@
 					   void main () { }
 	], [
 	   AC_MSG_RESULT(yes)
+	   # Pass it to libgc as well
+	   CFLAGS_FOR_LIBGC="$CFLAGS_FOR_LIBGC -mno-tls-direct-seg-refs"
 	], [
 	   AC_MSG_RESULT(no)
 	   CFLAGS=$ORIG_CFLAGS
@@ -1981,7 +2001,10 @@
 		# We should use a separate variable for this to avoid passing useless and
 		# potentially problematic defines to libgc (like -D_FILE_OFFSET_BITS=64)
 		# This should be executed late so we pick up the final version of CPPFLAGS
-		ac_configure_args="$ac_configure_args --disable-embed-check --with-libgc-threads=$libgc_threads $libgc_configure_args \"CPPFLAGS=$CPPFLAGS\""
+		# The problem with this approach, is that during a reconfigure, the main
+		# configure scripts gets invoked with these arguments, so we use separate
+		# variables understood by libgc's configure to pass CPPFLAGS and CFLAGS.
+		ac_configure_args="$ac_configure_args --disable-embed-check --with-libgc-threads=$libgc_threads $libgc_configure_args \"CPPFLAGS_FOR_LIBGC=$CPPFLAGS\" \"CFLAGS_FOR_LIBGC=$CFLAGS_FOR_LIBGC\""
 		AC_CONFIG_SUBDIRS(libgc)
 		;;
 esac

--- mono/libgc/configure.in	2008-05-19 14:08:04 UTC (rev 103473)
+++ mono/libgc/configure.in	2008-05-19 14:30:46 UTC (rev 103474)
@@ -39,6 +39,15 @@
 
 . [$]{srcdir}/configure.host
 
+# We use a separate variable to pass down CPPFLAGS and CFLAGS from the main mono 
+# configure, because of autoconf brokeness
+if test "x$CPPFLAGS_FOR_LIBGC" != "x"; then
+   CPPFLAGS=$CPPFLAGS_FOR_LIBGC
+fi
+if test "x$CFLAGS_FOR_LIBGC" != "x"; then
+   CFLAGS=$CFLAGS_FOR_LIBGC
+fi
+
 GC_CFLAGS=${gc_cflags}
 AC_SUBST(GC_CFLAGS)
